File: k9mail/src/main/java/com/fsck/k9/message/extractors/TextPartFinder.java
package com.fsck.k9.message.extractors;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.fsck.k9.mail.Body;
import com.fsck.k9.mail.BodyPart;
import com.fsck.k9.mail.PartHeaderMetadata;
import com.fsck.k9.mail.Multipart;
import com.fsck.k9.mail.Part;


class TextPartFinder {
    @Nullable
    public Part findFirstTextPart(@NonNull Part part) {
        Body body = part.getBody();

        if (body instanceof Multipart) {
            Multipart multipart = (Multipart) body;
            if (PartHeaderMetadata.from(part).isMimeType("multipart/alternative")) {
                return findTextPartInMultipartAlternative(multipart);
            } else {
                return findTextPartInMultipart(multipart);
            }
        } else if (PartHeaderMetadata.from(part).isMimeTypeAnyOf("text/plain", "text/html")) {
            return part;
        }

        return null;
    }

    private Part findTextPartInMultipartAlternative(Multipart multipart) {
        Part htmlPart = null;

        for (BodyPart bodyPart : multipart.getBodyParts()) {
            PartHeaderMetadata fancyBodyPart = PartHeaderMetadata.from(bodyPart);
            Body body = bodyPart.getBody();

            if (body instanceof Multipart) {
                Part candidatePart = findFirstTextPart(bodyPart);
                if (candidatePart != null) {
                    if (PartHeaderMetadata.from(candidatePart).isMimeType("text/html")) {
                        htmlPart = candidatePart;
                    } else {
                        return candidatePart;
                    }
                }
            } else if (fancyBodyPart.isMimeType("text/plain")) {
                return bodyPart;
            } else if (fancyBodyPart.isMimeType("text/html") && htmlPart == null) {
                htmlPart = bodyPart;
            }
        }

        if (htmlPart != null) {
            return htmlPart;
        }

        return null;
    }

    private Part findTextPartInMultipart(Multipart multipart) {
        for (BodyPart bodyPart : multipart.getBodyParts()) {
            Body body = bodyPart.getBody();

            if (body instanceof Multipart) {
                Part candidatePart = findFirstTextPart(bodyPart);
                if (candidatePart != null) {
                    return candidatePart;
                }
            } else if (PartHeaderMetadata.from(bodyPart).isMimeTypeAnyOf("text/plain", "text/html")) {
                return bodyPart;
            }
        }

        return null;
    }
}

File: app/core/src/main/java/com/fsck/k9/message/extractors/TextPartFinder.kt
package com.fsck.k9.message.extractors

import com.fsck.k9.mail.Multipart
import com.fsck.k9.mail.Part
import com.fsck.k9.mail.internet.MimeUtility.isSameMimeType

class TextPartFinder {
    fun findFirstTextPart(part: Part): Part? {
        val mimeType = part.mimeType
        val body = part.body

        return if (body is Multipart) {
            if (isSameMimeType(mimeType, "multipart/alternative")) {
                findTextPartInMultipartAlternative(body)
            } else {
                findTextPartInMultipart(body)
            }
        } else if (isSameMimeType(mimeType, "text/plain") || isSameMimeType(mimeType, "text/html")) {
            part
        } else {
            null
        }
    }

    private fun findTextPartInMultipartAlternative(multipart: Multipart): Part? {
        var htmlPart: Part? = null

        for (bodyPart in multipart.bodyParts) {
            val mimeType = bodyPart.mimeType
            val body = bodyPart.body

            if (body is Multipart) {
                val candidatePart = findFirstTextPart(bodyPart)
                if (candidatePart != null) {
                    htmlPart = if (isSameMimeType(candidatePart.mimeType, "text/html")) {
                        candidatePart
                    } else {
                        return candidatePart
                    }
                }
            } else if (isSameMimeType(mimeType, "text/plain")) {
                return bodyPart
            } else if (isSameMimeType(mimeType, "text/html") && htmlPart == null) {
                htmlPart = bodyPart
            }
        }

        return htmlPart
    }

    private fun findTextPartInMultipart(multipart: Multipart): Part? {
        for (bodyPart in multipart.bodyParts) {
            val mimeType = bodyPart.mimeType
            val body = bodyPart.body

            if (body is Multipart) {
                val candidatePart = findFirstTextPart(bodyPart)
                if (candidatePart != null) {
                    return candidatePart
                }
            } else if (isSameMimeType(mimeType, "text/plain") || isSameMimeType(mimeType, "text/html")) {
                return bodyPart
            }
        }

        return null
    }
}
